test/spectests/spectests.cpp

```diff
@@ -420,26 +403,57 @@ class test_runner
         return it_instance->second.get();
     }
 
-    std::optional<fizzy::Value> read_value(const json& v)
+
+    static fizzy::Value read_integer_value(const json& v)
     {
-        const auto type = v.at("type").get<std::string>();
+        assert(
+            v.at("type").get<std::string>() == "i32" || v.at("type").get<std::string>() == "i64");
 
         // JSON tests have all values including floats serialized as 64-bit unsigned integers
+        return std::stoull(v.at("value").get<std::string>());
+    }
+
+    template <typename T>
+    static fizzy::Value read_floating_point_value(const json& v)
+    {
+        assert(
+            v.at("type").get<std::string>() == "f32" || v.at("type").get<std::string>() == "f64");
+        assert(!is_canonical_nan(v) && !is_arithmetic_nan(v));
+
+        // JSON tests have all values including floats serialized as 64-bit unsigned integers.
         const uint64_t uint_value = std::stoull(v.at("value").get<std::string>());
```

Comment:
@gumb0 Did not mention before merging, because this question affected `master` too (e.g. not introduced via this PR): do we want to wrap any of these in proper exception handling for invalid inputs (see https://en.cppreference.com/w/cpp/string/basic_string/stoul for `invalid_argument` or `out_of_range`) or it is fine leaving it to the general exception catching we have?

Reply:
Actually it would be nice to output a line number in the case of exception, now you see only that exception happened somewhere in the file.

Reply:
Wait until it happens again.